tests/heuristic_calc_test.pl

```diff
@@ -0,0 +1,57 @@
+:- consult('../alphabeta/heuristic_calc.pl'). % import possible states impl.
+
+run:-
+    % test numberOfPawns
+
+    numberOfPawns(white, [piece(3,3,black,false),piece(4,4,black,false)], 0),
+    numberOfPawns(white, [piece(3,3,white,false),piece(4,4,black,false)], 1),
+    numberOfPawns(white, [piece(3,3,white,false),piece(4,4,white,true)], 1),
+    numberOfPawns(white, [piece(3,3,white,true),piece(4,4,white,true)], 0),
+    writeln('numberOfPawns - done!'),
+
+    numberOfQueens(white, [piece(3,3,black,false),piece(4,4,black,true)], 0),
+    numberOfQueens(white, [piece(3,3,white,false),piece(4,4,white,false)], 0),
+    numberOfQueens(white, [piece(3,3,white,false),piece(4,4,white,true)], 1),
+    numberOfQueens(white, [piece(3,3,white,true),piece(4,4,white,true)], 2),
+    writeln('numberOfQueens - done!'),
+
+    numberOfSafePawns(white, [piece(0,0,black,false),piece(5,4,black,true)], 6, 0),
+    numberOfSafePawns(white, [piece(0,0,white,false),piece(5,4,white,false)], 6, 2),
+    numberOfSafePawns(white, [piece(0,0,white,false),piece(4,4,white,false)], 5, 2),
+    numberOfSafePawns(white, [piece(0,0,white,false),piece(4,4,white,true)], 5, 1),
+    writeln('numberOfSafePawns - done!'),
+
+    numberOfSafeQueens(white, [piece(0,0,black,false),piece(5,4,black,true)], 6, 0),
+    numberOfSafeQueens(white, [piece(0,0,white,true),piece(5,4,white,false)], 6, 1),
+    numberOfSafeQueens(white, [piece(0,0,white,true),piece(4,4,white,true)], 5, 2),
+    numberOfSafeQueens(white, [piece(0,0,white,false),piece(4,4,white,true)], 5, 1),
+    writeln('numberOfSafeQueens - done!'),
+
+    aggregatedDistanceOfPawnsToPromotionLine(white, [piece(0,0,white,false),piece(2,2,white,true)], 5, 4),
+    aggregatedDistanceOfPawnsToPromotionLine(black, [piece(4,4,black,false),piece(2,2,black,true)], 5, 4),
+    aggregatedDistanceOfPawnsToPromotionLine(white, [piece(0,0,white,true),piece(4,4,white,true)], 5, 0),
+
+    writeln('aggregatedDistanceOfPawnsToPromotionLine - done!'),
+
+    numberOfUnoccupiedFieldsOnPromotionLine(white, [piece(4,0,white,false),piece(4,2,white,true)], 4, 2),
+    numberOfUnoccupiedFieldsOnPromotionLine(black, [piece(0,4,black,false),piece(0,2,black,true)], 4, 2),
+    numberOfUnoccupiedFieldsOnPromotionLine(black, [piece(0,4,black,false), piece(0,1,black,false)], 4, 2),
+    numberOfUnoccupiedFieldsOnPromotionLine(black, 
+        [piece(0,4,white,false), piece(0,4,white,false), piece(0,4,white,false),piece(0,2,black,true)]
+        , 8, 1),
+    numberOfUnoccupiedFieldsOnPromotionLine(white, 
+            [piece(7,4,black,false), piece(7,2,black,false), piece(7,4,white,false),piece(7,2,black,true)]
+            , 8, 1),
+    numberOfUnoccupiedFieldsOnPromotionLine(white, 
+        [piece(7,4,black,false), piece(7,4,black,false), piece(7,4,black,false),piece(7,2,black,true)]
```

Comment:
why 0 is available? 
7,0
7,6
are available

Reply:
same as before:
the occupied fields is counting rival piece on promotion line, and the final result is a sub between the number of available tile (black tile) to rival piece on promotion line. The column index means nothing. (The game client won't send a board with 2 pieces on the same tile)